engine/execution/storehouse.go

```diff
@@ -0,0 +1,74 @@
+package execution
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// RegisterStore is the interface for register store
+// see implementation in engine/execution/storehouse/register_store.go
+type RegisterStore interface {
+	// GetRegister first try to get the register from InMemoryRegisterStore, then OnDiskRegisterStore
+	GetRegister(height uint64, blockID flow.Identifier, register flow.RegisterID) (flow.RegisterValue, error)
+
+	// SaveRegisters saves to InMemoryRegisterStore first, then trigger the same check as OnBlockFinalized
+	// Depend on InMemoryRegisterStore.SaveRegisters
+	SaveRegisters(header *flow.Header, registers []flow.RegisterEntry) error
+
+	// Depend on FinalizedReader's GetFinalizedBlockIDAtHeight
+	// Depend on ExecutedFinalizedWAL.Append
+	// Depend on OnDiskRegisterStore.SaveRegisters
+	// OnBlockFinalized trigger the check of whether a block at the next height becomes finalized and executed.
+	// the next height is the existing finalized and executed block's height + 1.
+	// If a block at next height becomes finalized and executed, then:
+	// 1. write the registers to write ahead logs
+	// 2. save the registers of the block to OnDiskRegisterStore
+	// 3. prune the height in InMemoryRegisterStore
+	OnBlockFinalized() error
```

Comment:
to reiterate the conversation from the offline discussion, can this implement `OnFinalizedBlock(*model.Block)` so it can work with `*pubsub.FollowerDistributor`? That would make it better align with other components.

It would have to be made non-blocking, but there are existing patterns to solve that.

Reply:
You are right, this method is better to be non-blocking. However, if the pruning is done in a separate goroutine, we need to be able to handle any potential error, which means, we need to make it stateful, just like `ExecutionDataRequester`.

I actually would like to keep this module simple, because if OnBlockFInalized is blocking, it's much easier to test compared to non-blocking where the work is done is separate goroutine.

In order to make this method non-blocking, I can create another module which wraps the RegisterStore, using the engine.Notifier to implement non-blocking `OnFinalizedBlock` method, and implement the stateful component API to be initialized. So the wrapper just need to focus on dealing with the non-blocking call, as well as handling errors from the goroutine.